[3.x] Add a Preview Bake button for quick iteration with BakedLightmap

Calinou opened the pull request:
- `3.x` version of https://github.com/godotengine/godot/pull/80518.

When using the **Preview Bake** button instead of **Bake Lightmaps**, low quality settings will be used instead of settings from the BakedLightmap node. A node configuration warning will be displayed besides the BakedLightmap node to inform the user that the bake was done in preview mode. This preview bake warning is persisted to storage to make teamwork easier.

![Editor top bar](https://user-images.githubusercontent.com/180032/104819367-48051f80-582d-11eb-988a-dc40022c90fe.png)

![Node configuration warning](https://user-images.githubusercontent.com/180032/104820646-a1258100-5836-11eb-8a99-c6a81a42dbeb.png)

This helps iterate quicker on a scene's lighting since you don't need to wait as much time to see results.

On a [simple test project](https://github.com/godotengine/godot/files/5824856/test_lightmap.zip), Preview Bake (0:00:38) was about 3 times faster than a "final" bake (0:01:48) with the default BakedLightmap settings.


By default, Preview Bake uses:

- Low quality for baking.
- 1 bounce, unless the BakedLightmap node is set to 0 bounces (in which case it uses 0 bounces as well).
- Low quality for dynamic object capture.
  - You can disable dynamic object capture entirely in the Project Settings if you don't need it for previewing purposes.

The preview bake settings above can be adjusted in the Project Settings. These are project settings so that different team members always use the same preview settings for consistency.

[**Comparison between Preview Bake and "final" bake**](https://imgsli.com/Mzc4NDA)

- This addresses https://github.com/godotengine/godot-proposals/issues/2029 for the `3.x` branch.

Calinou committed:
Add a Preview Bake button for quick iteration with BakedLightmap

When using the Preview Bake button instead of Bake Lightmaps,
low quality settings will be used instead of settings from the
BakedLightmap node. A node configuration warning will be displayed
besides the BakedLightmap node to inform the user that the bake
was done in preview mode. This preview bake warning is persisted
to storage to make teamwork easier.

This helps iterate quicker on a scene's lighting since you don't
need to wait as much time to see results.

On a simple test scene, Preview Bake was about 3 times faster
than a "final" bake with the default BakedLightmap settings.

The preview bake settings can be adjusted in the Project Settings.

Zireael07 commented:
If I understand it correctly, I *can* use preview bake if I want to in production (because real bakes will likely take too long on my hardware), this is just a suggestion/reminder?

Calinou commented:
> If I understand it correctly, I can use preview bake if I want to in production (because real bakes will likely take too long on my hardware), this is just a suggestion/reminder?

Yes, the bake is still saved and can be used in an exported project. That said, I'd recommend you change the BakedLightmap node properties instead of relying on Preview Bake.

JFonS commented on scene/3d/baked_lightmap.cpp:
```diff
@@ -1238,6 +1272,15 @@ BakedLightmap::BakeQuality BakedLightmap::get_capture_quality() const {
 	return capture_quality;
 }
 
+void BakedLightmap::_set_preview_bake(bool p_preview_bake) {
+	preview_bake = p_preview_bake;
+	update_configuration_warning();
+}
+
+bool BakedLightmap::_is_preview_bake() const {
+	return preview_bake;
+}
+
```
I would move the "preview_bake" flag to `BakedLightmapData`. I think it makes more sense to mark the resource as "baked in preview mode", rather than setting the flag in the node itself.  Since lightmap data is saved in files by default, a user could load a different `BakedLightmapData` resource and the node would still be flagged as using a preview bake.

Maybe we could even add a `.preview.` infix to the saved file name, so baking in preview mode doesn't overwrite a "final" bake you may already have.

Calinou commented on scene/3d/baked_lightmap.cpp:
```diff
@@ -775,7 +805,7 @@ BakedLightmap::BakeError BakedLightmap::bake(Node *p_from_node, String p_data_sa
 	}
 
-	if (capture_enabled) {
+	if (capture_enabled && bool(GLOBAL_GET("rendering/cpu_lightmapper/preview_bake/allow_capture"))) {
```
Indeed, I need to change this so the preview bake setting only applies for preview bakes.

Calinou commented:
I rebased and moved the `preview_bake` property to BakedLightmapData as requested by @JFonS. I tested it again, it works as expected :slightly_smiling_face: 

> Maybe we could even add a .preview. infix to the saved file name, so baking in preview mode doesn't overwrite a "final" bake you may already have.

I'm not sure how to handle that part without accidentally exporting those preview bakes in the final PCK (if they weren't removed by the user). I think preview bake data could even be stored in `.godot/`, but I'd leave that for a future PR.
